print-geonet.c

```diff
@@ -26,232 +27,481 @@
+	{2009, "CPM"},
+	{2010, "EVCSN_POI"},
+	{2011, "TPG"},
+	{2012, "EV_RSR"},
+	{2013, "RTCMEM"},
+	{2014, "CTLM"},
+	{2015, "CRLM"},
+	{2016, "EC_AT_REQ"},
+	{2017, "MCDM"},
+	{2018, "VAM"},
+	{2019, "IMZM"},
+	{2020, "DSM"},
+	{2021, "P2P_CRLM"},
+	{2022, "P2P_CTLM"},
+	{2023, "MRS"},
+	{2024, "P2P_FULL_CTLM"},
+	{0, NULL}};
+
 
-	/* Assuming ItsPduHeader */
-	ND_PRINT("; ItsPduHeader v:%u", GET_U_1(bp));
 
-	msg_type = GET_U_1(bp + 1);
-	ND_PRINT(" t:%u-%s", msg_type,
-	         tok2str(msg_type_values, "unknown (%u)", msg_type));
+static int is_value_in_list(u_int value, const u_int*list, u_int length_of_list){
+	for (u_int i = 0; i < length_of_list; i++){
+		if (value == list[i]){
+			return 1;
+		}
+	}
+	return 0;
 }
 
-/* EN 302 636-5-1 V2.2.1 Section 7.2: BTP-A header */
-static void
-print_btp(netdissect_options *ndo,
-	  const u_char *bp)
+
+static u_int convert_lt_to_seconds(u_int lt_base, u_int lt_multiplier)
 {
-	ND_PRINT("; BTP Dst:%u", GET_BE_U_2(bp + 0));
-	ND_PRINT(" Src:%u", GET_BE_U_2(bp + 2));
+	float base_seconds;
+	switch (lt_base)
+	{
+	case 0: // 50 milliseconds
+		base_seconds = 0.05;
+		break;
+	case 1: // 1 second
+		base_seconds = 1.0;
+		break;
+	case 2: // 10 seconds
+		base_seconds = 10.0;
+		break;
+	case 3: // 100 seconds
+		base_seconds = 100.0;
+		break;
+	default: // default to 0 second
+		base_seconds = 0.0;
+		break;
+	}
+	return (u_int)(base_seconds * lt_multiplier);
 }
 
-static void
-print_long_pos_vector(netdissect_options *ndo,
-		      const u_char *bp)
+static void gn_basic_header_decode_from_bytes(netdissect_options *ndo, const u_char **bp, u_int *length, u_int *next_header)
 {
-	ND_PRINT("GN_ADDR:%s ", GET_LINKADDR_STRING(bp, LINKADDR_OTHER, GEONET_ADDR_LEN));
-	ND_PRINT("lat:%u ", GET_BE_U_4(bp + 12));
-	ND_PRINT("lon:%u", GET_BE_U_4(bp + 16));
+	u_int version;
+	u_int reserved;
+	u_int lt_multiplier;
+	u_int lt_base;
+	u_int rhl;
+
+	uint32_t value = GET_BE_U_4(*bp);
+	*bp += 4;
+	*length -= 4;
+	version = (value >> (4 + THREE_BYTES)) & FOUR_BITS_MASK;
+	if (!is_value_in_list(version, implemented_gn_versions, IMPLEMENTED_GN_VERSIONS_NUM))
+	{
+		ND_PRINT(" (Unsupported GeoNetworking Basic Header version %u)", version);
+		*next_header = 0; // Indicates an error.
+		return;
+	}
+	*next_header = (value >> (THREE_BYTES)) & FOUR_BITS_MASK;
+	reserved = (value >> (TWO_BYTES)) & EIGHT_BITS_MASK;
+	lt_multiplier = (value >> (2 + ONE_BYTE)) & SIX_BITS_MASK;
+	lt_base = (value >> ONE_BYTE) & THREE_BITS_MASK;
+	rhl = value & FOUR_BITS_MASK;
+
+	const char *next_header_text = tok2str(basic_header_next_header_values, "Unknown", *next_header);
+	u_int lt_seconds = convert_lt_to_seconds(lt_base, lt_multiplier);
+	if (ndo->ndo_vflag == NDO_V_FLAG_FIRST_DEBUG_LEVEL)
+	{
+		ND_PRINT("ver:%u nh:%s lt:%us rhl:%u; ",
+				 version, next_header_text, lt_seconds, rhl);
+	}
+	else if (ndo->ndo_vflag > NDO_V_FLAG_FIRST_DEBUG_LEVEL)
+	{
+		ND_PRINT("ver:%u nh:%s reserved:%u lt:[base:%u mult:%u = %us] rhl:%u; ",
```

Comment:
And this. Also here LT base could be decoded using `tok2str()`. Also it may be simpler to print LT using the same format in all cases, unless there is a good reason to make the verbosity granular.